File: alarmeitbl/comandos.py
#!/usr/bin/env python3

import time
from .myeventloop.tcpclient import *
from .utils_proto import *

# Envio de comandos diretamente do cliente para a central de alarme
# usando o novo protocolo ISECNet2 (o mesmo do download de fotos)

class ComandarCentral(TCPClientHandler, UtilsProtocolo):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__((ip_addr, cport))
        self.log_debug("Inicio")
        self.observer = observer
        self.conn_timeout = self.timeout("conn_timeout", 15, self.conn_timeout)
        self.senha = senha
        self.tam_senha = tam_senha
        self.extra = extra
        self.status = 1
        self.tratador = None

    def destroyed_callback(self):
        if self.observer:
            self.observer.resultado(self.status)

    def conn_timeout(self, task):
        self.log_info("Timeout")
        if self.status != 0:
            # se status == 0, provavelmente já completou a tarefa e o timeout é na despedida
            self.status = 1
        self.destroy()

    def connection_callback(self, ok):
        self.conn_timeout.cancel()
        if not ok:
            self.status = 1
            self.log_info("Conexao falhou")
            # destroy() executado pelo chamador
            return
        self.autenticacao()

    def autenticacao(self):
        self.log_debug("Autenticacao")
        pct = self.pacote_isecnet2_auth(self.senha, self.tam_senha)
        self.log_debug("Send", self.hexprint(pct))
        self.send(pct)
        self.tratador = self.resposta_autenticacao
        self.conn_timeout.restart()

    def recv_callback(self, latest):
        self.log_debug("Recv", self.hexprint(latest))

        compr = self.pacote_isecnet2_completo(self.recv_buf)
        if not compr:
            self.log_debug("Pacote incompleto")
            return

        pct, self.recv_buf = self.recv_buf[:compr], self.recv_buf[compr:]

        if not self.pacote_isecnet2_correto(pct):
            self.log_info("Pacote incorreto, desistindo")
            self.destroy()
            return

        cmd, payload = self.pacote_isecnet2_parse(pct)
        self.log_debug("Resposta %04x" % cmd)

        if not self.tratador:
            self.log_info("Sem tratador")
            self.destroy()
            return

        self.conn_timeout.cancel()
        self.tratador(cmd, payload)

    def resposta_autenticacao(self, cmd, payload):
        if cmd == 0xf0fd:
            self.nak(payload)
            return

        if cmd != 0xf0f0:
            self.log_info("Autenticacao: resp inesperada %04x" % cmd)
            self.destroy()
            return

        if len(payload) != 1:
            self.log_info("Autenticacao: resposta invalida")
            self.destroy()
            return

        resposta = payload[0]
        # Possíveis respostas:
        # 01 = senha incorreta
        # 02 = versão software incorreta
        # 03 = painel chamará de volta (?)
        # 04 = aguardando permissão de usuário (?)
        if resposta > 0:
            self.log_info("Autenticacao: falhou motivo %d" % resposta)
            self.destroy()
            return

        self.log_debug("Autenticacao ok")
        self.envia_comando_in()

    def envia_comando(self, cmd, payload, tratador_in):
        pct = self.pacote_isecnet2(cmd, payload)
        self.log_debug("Send", self.hexprint(pct))
        self.send(pct)

        self.cmd = cmd
        self.tratador = self.resposta_comando
        self.tratador_in = tratador_in

        self.conn_timeout.restart()

    def resposta_comando(self, cmd, payload):
        if cmd == 0xf0fd:
            self.nak(payload)
            return

        if cmd == 0xf0f7:
            self.log_info("Erro central ocupada")
            self.destroy()
            return

        if cmd != self.cmd and cmd != 0xf0fe:
            self.log_info("Resposta inesperada %04x" % cmd)
            self.destroy()
            return

        self.tratador_in(payload)

    def despedida(self):
        self.log_debug("Despedindo")
        pct = self.pacote_isecnet2_bye()
        self.log_debug("Send", self.hexprint(pct))
        self.send(pct)

        self.tratador = None
        # Reportar sucesso ao observador
        self.status = 0
        self.conn_timeout.restart()
        # Resposta esperada: central fechar conexão

    def nak(self, payload):
        if len(payload) != 1:
            self.log_info("NAK invalido")
        else:
            motivo = payload[0]
            self.log_info("NAK motivo %02x" % motivo)
            self.status = 1
        self.destroy()


class AtivarDesativarCentral(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra, subcmd):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)
        self.particao = extra[0]
        self.subcmd = subcmd

    def envia_comando_in(self):
        # byte 1: particao (0x01 = 1, 0xff = todas ou sem particao)
        # byte 2: 0x00 desarmar, 0x01 armar, 0x02 stay
        if self.particao is None:
            payload = [ 0xff, self.subcmd ]
        else:
            payload = [ self.particao, self.subcmd ]
        self.envia_comando(0x401e, payload, self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        self.despedida()

class DesativarCentral(AtivarDesativarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra, 0x00)

class AtivarCentral(AtivarDesativarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra, 0x01)


class DesligarSirene(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)
        self.particao = extra[0]

    def envia_comando_in(self):
        if self.particao is None:
            payload = [ 0xff ]
        else:
            payload = [ self.particao]
        self.envia_comando(0x4019, payload, self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        self.despedida()


class LimparDisparo(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)

    def envia_comando_in(self):
        self.envia_comando(0x4013, [], self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        self.despedida()


class CancelarZona(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)
        self.zona = extra[0]

    def envia_comando_in(self):
        # TODO Suportar todas as zonas (enviar 0xff)
        if not self.zona or self.zona < 1 or self.zona > 254:
            raise Exception("Zona precisa ser especificada.")
        payload = [ self.zona - 1, 0x01 ]
        self.envia_comando(0x401f, payload, self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        self.despedida()

class ReativarZona(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)
        self.zona = extra[0]

    def envia_comando_in(self):
        # TODO Suportar todas as zonas (enviar 0xff)
        if not self.zona or self.zona < 1 or self.zona > 254:
            raise Exception("Zona precisa ser especificada.")
        payload = [ self.zona - 1, 0x00 ]
        self.envia_comando(0x401f, payload, self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        self.despedida()


class SolicitarStatus(ComandarCentral):
    def __init__(self, observer, ip_addr, cport, senha, tam_senha, extra):
        super().__init__(observer, ip_addr, cport, senha, tam_senha, extra)

    def envia_comando_in(self):
        self.envia_comando(0x0b4a, [], self.resposta_comando_in)

    def resposta_comando_in(self, payload):
        # Documentação é base 1
        payload = [0] + payload
        print()
        print("*******************************************")
        if payload[1] == 0x01:
            print("Central AMT-8000")
        else:
            print("Central de tipo desconhecido")
        print("Versão de firmware %d.%d.%d" % tuple(payload[2:5]))
        print("Status geral: ")
        armado = {0x00: "Desarmado", 0x01: "Partição(ões) armada(s)", 0x03: "Todas partições armadas"}
        print("\t" + armado[((payload[21] >> 5) & 0x03)])
        print("\tZonas em alarme:", (payload[21] & 0x8) and "Sim" or "Não")
        print("\tZonas canceladas:", (payload[21] & 0x10) and "Sim" or "Não")
        print("\tTodas zonas fechadas:", (payload[21] & 0x4) and "Sim" or "Não")
        print("\tSirene:", (payload[21] & 0x2) and "Sim" or "Não")
        print("\tProblemas:", (payload[21] & 0x1) and "Sim" or "Não")
        for particao in range(0, 17):
            habilitado = payload[22 + particao] & 0x80
            if not habilitado:
                continue
            print("Partição %02d:" % particao)
            print("\tStay:", (payload[22 + particao] & 0x40) and "Sim" or "Não")
            print("\tDelay de saída:", (payload[22 + particao] & 0x20) and "Sim" or "Não")
            print("\tPronto para armar:", (payload[22 + particao] & 0x10) and "Sim" or "Não")
            print("\tAlame ocorreu:", (payload[22 + particao] & 0x08) and "Sim" or "Não")
            print("\tEm alarme:", (payload[22 + particao] & 0x04) and "Sim" or "Não")
            print("\tArmado modo stay:", (payload[22 + particao] & 0x02) and "Sim" or "Não")
            print("\tArmado:", (payload[22 + particao] & 0x01) and "Sim" or "Não")
        print("Zonas abertas:", self.bits_para_numeros(payload[39:47]))
        print("Zonas em alarme:", self.bits_para_numeros(payload[47:55]))
        # print("Zonas ativas:", self.bits_para_numeros(payload[55:63], inverso=True))
        print("Zonas em bypass:", self.bits_para_numeros(payload[55:63]))
        print("Sirenes ligadas:", self.bits_para_numeros(payload[63:65]))

        # TODO interpretar mais campos
        print("*******************************************")
        print()

        self.despedida()

    def bits_para_numeros(self, octetos, inverso=False):
        lista = []
        for i, octeto in enumerate(octetos):
            for j in range(0, 8):
                bit = octeto & (1 << j)
                if (bit and not inverso) or (not bit and inverso):
                    lista.append("%d" % (1 + j + i * 8))
        return ", ".join(lista)

File: goalarmeitbl/comandos.go
package goalarmeitbl

import (
    "slices"
    "strings"
    "fmt"
)

// Construtor de uma implementação/subclasse
type Constructor func(int) (ComandoCentralSub, string)

// Descritor de uma subclasse, para usar num mapa string -> descritor
type DescComandoSub struct {
    ExtraHelp string
    ExtraParam bool
    Construtor Constructor
}

// Apenas autentica e encerra.
// Útil para testes, conferir que a senha é válida, etc.
type ComandoNulo struct {
}

func (comando *ComandoNulo) Autenticado(super *ComandoCentral) {
    super.Despedida()
}

func NewComandoNulo(_ int) (ComandoCentralSub, string) {
    comando := new(ComandoNulo)
    return comando, ""
}

// Solicita status da central: partições, disparos, etc.
type SolicitarStatus struct {
}

func (comando *SolicitarStatus) Autenticado(super *ComandoCentral) {
    pacote := PacoteIsecNet2(0x0b4a, nil)
    super.EnviarPacote(pacote, comando.RespostaStatus)
}

func bits_para_numeros(octetos []byte, inverso bool) string {
    lista := []string{}
    for i, octeto := range octetos {
        for j := range 8 {
            bit := (octeto & (1 << j))
            if (bit != 0 && !inverso) || (bit == 0 && inverso) {
                lista = append(lista, fmt.Sprintf("%d", (1 + j + i * 8)))
            }
        }
    }
    return strings.Join(lista, ", ")
}

func sim_nao(valor int) string {
    if valor != 0 {
        return "Sim"
    }
    return "Não"
}

func (comando *SolicitarStatus) RespostaStatus(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0x0b4a {
        fmt.Printf("RespostaStatus: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    payload = slices.Concat([]byte{0x00}, payload)
    fmt.Println()
    fmt.Println()
    fmt.Println("*******************************************")
    if payload[1] == 0x01 {
        fmt.Println("Central AMT-8000")
    } else {
        fmt.Println("Central de tipo desconhecido")
    }
    fmt.Printf("Versão de firmware %d.%d.%d\n", payload[2], payload[3], payload[4])
    fmt.Println("Status geral: ")
    var armado = map[int]string{0x00: "Desarmado", 0x01: "Partição(ões) armada(s)", 0x03: "Todas partições armadas"}
    fmt.Printf("\t %s\n", armado[int(((payload[21] >> 5) & 0x03))])
    fmt.Printf("\tZonas em alarme: %s\n", sim_nao(int(payload[21] & 0x8)))
    fmt.Printf("\tZonas canceladas: %s\n", sim_nao(int(payload[21] & 0x10)))
    fmt.Printf("\tTodas zonas fechadas: %s\n", sim_nao(int(payload[21] & 0x4)))
    fmt.Printf("\tSirene: %s\n", sim_nao(int(payload[21] & 0x2)))
    fmt.Printf("\tProblemas: %s\n", sim_nao(int(payload[21] & 0x1)))
    for particao := range 17 {
        habilitado := payload[22 + particao] & 0x80
        if habilitado == 0 {
            continue
        }
        fmt.Printf("Partição %02d:\n", particao)
        fmt.Printf("\tStay: %s\n", sim_nao(int(payload[22 + particao] & 0x40)))
        fmt.Printf("\tDelay de saída: %s\n", sim_nao(int(payload[22 + particao] & 0x20)))
        fmt.Printf("\tPronto para armar: %s\n", sim_nao(int(payload[22 + particao] & 0x10)))
        fmt.Printf("\tAlame ocorreu: %s\n", sim_nao(int(payload[22 + particao] & 0x08)))
        fmt.Printf("\tEm alarme: %s\n", sim_nao(int(payload[22 + particao] & 0x04)))
        fmt.Printf("\tArmado modo stay: %s\n", sim_nao(int(payload[22 + particao] & 0x02)))
        fmt.Printf("\tArmado: %s\n", sim_nao(int(payload[22 + particao] & 0x01)))
    }
    fmt.Printf("Zonas abertas: %s\n", bits_para_numeros(payload[39:47], false))
    fmt.Printf("Zonas em alarme: %s\n", bits_para_numeros(payload[47:55], false))
    // fmt.Printf("Zonas ativas: %s\n", bits_para_numeros(payload[55:63], true))
    fmt.Printf("Zonas em bypass: %s\n", bits_para_numeros(payload[55:63], false))
    fmt.Printf("Sirenes ligadas: %s\n", bits_para_numeros(payload[63:65], false))

    // TODO interpretar mais campos
    fmt.Println("*******************************************")
    fmt.Println()

    super.Despedida()
}

func NewSolicitarStatus(_ int) (ComandoCentralSub, string) {
    comando := new(SolicitarStatus)
    return comando, ""
}

// Desarmar o alarme da central
type DesativarCentral struct {
    particao int
}

func (comando *DesativarCentral) Autenticado(super *ComandoCentral) {
    // byte 1: particao (0x01 = 1, 0xff = todas ou sem particao)
    // byte 2: 0x00 desarmar, 0x01 armar, 0x02 stay
    pacote := PacoteIsecNet2(0x401e, []byte{byte(comando.particao), 0x00})
    super.EnviarPacote(pacote, comando.RespostaDesativarCentral)
}

func (comando *DesativarCentral) RespostaDesativarCentral(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0x401e {
        fmt.Printf("DesativarCentral: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewDesativarCentral(particao int) (ComandoCentralSub, string) {
    comando := new(DesativarCentral)
    if particao == 0 {
        // todas as partições
        comando.particao = 0xff
    } else {
        comando.particao = particao
    }
    return comando, ""
}

// Ativar o alarme, ou seja, armar a central
type AtivarCentral struct {
    particao int
}

func (comando *AtivarCentral) Autenticado(super *ComandoCentral) {
    // byte 1: particao (0x01 = 1, 0xff = todas ou sem particao)
    // byte 2: 0x00 desarmar, 0x01 armar, 0x02 stay
    pacote := PacoteIsecNet2(0x401e, []byte{byte(comando.particao), 0x01})
    super.EnviarPacote(pacote, comando.RespostaAtivarCentral)
}

func (comando *AtivarCentral) RespostaAtivarCentral(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0x401e {
        fmt.Printf("AtivarCentral: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewAtivarCentral(particao int) (ComandoCentralSub, string) {
    comando := new(AtivarCentral)
    if particao == 0 {
        // todas as partições
        comando.particao = 0xff
    } else {
        comando.particao = particao
    }
    return comando, ""
}

// Desligar a sirene, sem limpar o disparo do alarme em si
type DesligarSirene struct {
    particao byte
}

func (comando *DesligarSirene) Autenticado(super *ComandoCentral) {
    pacote := PacoteIsecNet2(0x4019, []byte{comando.particao})
    super.EnviarPacote(pacote, comando.RespostaDesligarSirene)
}

func (comando *DesligarSirene) RespostaDesligarSirene(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0xf0fe {
        fmt.Printf("DesligarSirene: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewDesligarSirene(particao int) (ComandoCentralSub, string) {
    comando := new(DesligarSirene)
    if particao == 0 {
        // todas as partições
        comando.particao = 0xff
    } else {
        comando.particao = byte(particao)
    }
    return comando, ""
}

// Bypass de zona, ou seja, desativar a zona de modo que não possa disparar o alarme
type BypassZona struct {
    zona byte
}

func (comando *BypassZona) Autenticado(super *ComandoCentral) {
    pacote := PacoteIsecNet2(0x401f, []byte{comando.zona - 1, 0x01})
    super.EnviarPacote(pacote, comando.RespostaBypassZona)
}

func (comando *BypassZona) RespostaBypassZona(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0xf0fe {
        fmt.Printf("BypassZona: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewBypassZona(zona int) (ComandoCentralSub, string) {
    comando := new(BypassZona)
    if zona < 1 || zona > 254 {
        return nil, "Zona precisa ser especificada e estar na faixa 1-254"
    }
    comando.zona = byte(zona)
    return comando, ""
}

// Reativar zona, ou seja, remover o bypass de zona
type ReativarZona struct {
    zona byte
}

func (comando *ReativarZona) Autenticado(super *ComandoCentral) {
    pacote := PacoteIsecNet2(0x401f, []byte{comando.zona - 1, 0x00})
    // TODO implementar reativação todas as zonas (0xff + códigos x 0x3f)
    super.EnviarPacote(pacote, comando.RespostaReativarZona)
}

func (comando *ReativarZona) RespostaReativarZona(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0xf0fe {
        fmt.Printf("ReativarZona: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewReativarZona(zona int) (ComandoCentralSub, string) {
    comando := new(ReativarZona)
    if zona < 1 || zona > 254 {
        return nil, "Zona precisa ser especificada e estar na faixa 1-254"
    }
    comando.zona = byte(zona)
    return comando, ""
}

// Limpar registro de alarme disparado
type LimparDisparo struct {
}

func (comando *LimparDisparo) Autenticado(super *ComandoCentral) {
    pacote := PacoteIsecNet2(0x4013, nil)
    super.EnviarPacote(pacote, comando.RespostaLimparDisparo)
}

func (comando *LimparDisparo) RespostaLimparDisparo(super *ComandoCentral, cmd int, payload []byte) {
    if cmd != 0xf0fe {
        fmt.Printf("LimparDisparo: resp inesperada %04x\n", cmd)
        super.Bye()
        return
    }

    super.Despedida()
}

func NewLimparDisparo(_ int) (ComandoCentralSub, string) {
    comando := new(LimparDisparo)
    return comando, ""
}

// Lista de comandos disponíveis

var Subcomandos map[string]DescComandoSub

func init() {
    Subcomandos = map[string]DescComandoSub{
        "nulo": DescComandoSub{"", false, NewComandoNulo},
        "status": DescComandoSub{"", false, NewSolicitarStatus},
        "ativar": DescComandoSub{"[partição] (se omitida, ativa todas)", true, NewAtivarCentral},
        "desativar": DescComandoSub{"[partição] (se omitida, desativa todas)", true, NewDesativarCentral},
        "desligarsirene": DescComandoSub{"[partição] (se omitida, desliga todas)", true, NewDesligarSirene},
        "limpardisparo": DescComandoSub{"", false, NewLimparDisparo},
        "bypass": DescComandoSub{"<zona> (obrigatório especificar zona)", true, NewBypassZona},
        "cancelbypass": DescComandoSub{"<zona> (obrigatório especificar zona)", true, NewReativarZona},
    }
}
